ada_feeding_perception/ada_feeding_perception/table_detection.py

```diff
@@ -0,0 +1,483 @@
+        ----------
+        camera_info: The camera information.
+        table_depth: The depth image of the table.
+
+        Returns
+        ----------
+        center: The center coordinates of the table in the camera frame.
+        q: The quaternion representing the orientation of the table plane.
+        """
+        # Deproject the depth image to get the 3D points in the camera frame.
+        pointcloud = depth_img_to_pointcloud(
+            table_depth,
+            0,
+            0,
+            f_x=camera_info.k[0],
+            f_y=camera_info.k[4],
+            c_x=camera_info.k[2],
+            c_y=camera_info.k[5],
+        )
+        xy_d = np.concatenate(
+            (pointcloud[:, 0, np.newaxis], pointcloud[:, 1, np.newaxis]), axis=1
+        )
+        depth_d = pointcloud[:, 2]
+        center = [
+            pointcloud[self.deproject_center_coord][0],
+            pointcloud[self.deproject_center_coord][1],
+            pointcloud[self.deproject_center_coord][2],
+        ]
+
+        # Fit Plane: z = a*x + b*y + c
+        coeffs = np.hstack((np.vstack(xy_d), np.ones((len(xy_d), 1)))).astype(float)
+        a, b, c = np.linalg.lstsq(coeffs, depth_d, rcond=None)[0]
+
+        # Modify the z coordinate of the center given the fitted plane (i.e. make the center
+        # the origin of the table plane)
+        center[2] = a * center[0] + b * center[1] + c
+
+        # Get the direction vectors of the table plane from the camera's frame of perspective
+        x_p = [
+            center[0] - self.dir_constant,
+            center[1],
+            a * (center[0] - self.dir_constant) + b * center[1] + c,
+        ]
+        x_ref = [center[0] - x_p[0], center[1] - x_p[1], center[2] - x_p[2]]
+        x_ref /= np.linalg.norm(x_ref)
+
+        z_ref = [a, b, -1] / np.linalg.norm([a, b, -1])
+
+        # Direction vector of y calculated through a system of linear equations consisting of:
+        # the dot product of x and y, dot product of z and y, & norm of y = 1
+        denom = math.sqrt(
+            math.pow(x_ref[2] * z_ref[1] - x_ref[1] * z_ref[2], 2)
+            + math.pow(x_ref[0] * z_ref[2] - x_ref[2] * z_ref[0], 2)
+            + math.pow(x_ref[1] * z_ref[0] - x_ref[0] * z_ref[1], 2)
+        )
+        y_ref = [
+            (x_ref[2] * z_ref[1] - x_ref[1] * z_ref[2]) / denom,
+            (x_ref[0] * z_ref[2] - x_ref[2] * z_ref[0]) / denom,
+            (x_ref[1] * z_ref[0] - x_ref[0] * z_ref[1]) / denom,
+        ]
+
+        # Construct rotation matrix from direction vectors
+        # TODO: Test tf_transformations quaternion_from_matrix function
+        rot_matrix = [
+            [x_ref[0], y_ref[0], z_ref[0]],
+            [x_ref[1], y_ref[1], z_ref[1]],
+            [x_ref[2], y_ref[2], z_ref[2]],
+        ]
+
+        # Derive quaternion from rotation matrix
+        q = R.from_matrix(rot_matrix)
+        q = q.as_quat()
+
+        return center, q
+
+    def toggle_table_detection_callback(
```

Comment:
This is a small nit but could you put these callbacks before the `fit_table` function? It improves readability for the functions to be ordered in some intuitive way, potentially in the way they are referred to. The callbacks are referred to from `__init__` whereas the other functions are referred to from `run`